### PR TITLE
 mon/nvmeofgw*: monitors publish in nvme-gw show ana group responsible

### PR DESCRIPTION
**Pull request by leonidc**
for  namespace rebalance

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/mon/NVMeofGwMon.cc [hunk 1]
```diff
@@ -66,6 +66,11 @@ void NVMeofGwMon::on_shutdown()
 
 void NVMeofGwMon::tick()
 {
+  if (++tick_ratio == 10) {
+    global_rebalance_index ++;
```
Comment (athanatos) [inline src/mon/NVMeofGwMon.cc:70]: global_rebalance_index++;

File: src/mon/NVMeofGwMon.cc [hunk 2]
```diff
@@ -317,6 +322,11 @@ bool NVMeofGwMon::preprocess_command(MonOpRequestRef op)
     f->dump_string("group", group);
     if (HAVE_FEATURE(mon.get_quorum_con_features(), NVMEOFHA)) {
       f->dump_string("features", "LB");
+      if (map.created_gws[group_key].size()) {
+        uint32_t index = (global_rebalance_index %
+              map.created_gws[group_key].size()) + 1;
+        f->dump_unsigned ("rebalance_ana_group", index);
```
Comment (athanatos) [inline src/mon/NVMeofGwMon.cc:328]: f->dump_unsigned("rebalance_ana_group", index);

File: src/mon/NVMeofGwMon.h [hunk 3]
```diff
@@ -83,6 +85,9 @@ class NVMeofGwMon: public PaxosService,
   void check_sub(Subscription *sub);
 
 private:
+  // used for calculate pool & group GW responsible for rebalance
+  uint32_t global_rebalance_index;
```
Comment (athanatos) [inline src/mon/NVMeofGwMon.h:89]: uint32_t global_rebalance_index = 1;
uint8_t  tick_ratio = 0;

File: src/mon/NVMeofGwMon.h [hunk 4]
```diff
@@ -45,6 +45,8 @@ class NVMeofGwMon: public PaxosService,
   NVMeofGwMon(Monitor &mn, Paxos &p, const std::string& service_name)
     : PaxosService(mn, p, service_name) {
     map.mon = &mn;
+    global_rebalance_index = 1;
+    tick_ratio = 0;
```
Comment (athanatos) [inline src/mon/NVMeofGwMon.h:49]: No need for this, initialize at the declaration site.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/mon/NVMeofGwMon.cc (hunk 2 above)
Comment (leonidc) [reply src/mon/NVMeofGwMon.cc:328]: sure

**Review by athanatos — APPROVED** [review]

**Comment by leonidc** [comment]
jenkins test api

**Pull request merged**